pre_commit/commands/run.py

```diff
@@ -417,10 +424,68 @@ def run(
         to_install = [hook for hook in hooks if hook.id not in skips]
         install_hook_envs(to_install, store)
 
-        retval = _run_hooks(config, hooks, skips, args, environ)
+        if args.hook_stage == 'commit' and not is_git_message_supplied():
+            # Allow user to enter commit message concurrently with running pre-commit hooks to lower
+            # wait times.
+
+            # Run git commands before starting hooks to avoid race conditions and git lock errors.
+            commit_message_template = get_commit_message_template()
+
+            with concurrent.futures.ThreadPoolExecutor(max_workers=2) as ex:
+                ex.submit(edit_commit_message, commit_message_template)
+                retval_future = ex.submit(_run_hooks, config, hooks, skips, args, environ)
+            retval = retval_future.result()
+        else:
+            retval = _run_hooks(config, hooks, skips, args, environ)
+
         trace.set_success(retval == 0)
         return retval
 
 
     # https://github.com/python/mypy/issues/7726
     raise AssertionError('unreachable')
+
+def get_global_git_editor() -> List[str]:
+    # The repo-local editor has been set to a special script. This gets the globally configured
+    # editor.
+    editor_str = subprocess.run(['git', 'var', 'GIT_EDITOR'], cwd='/', check=True, capture_output=True).stdout.decode('utf-8')
+    return shlex.split(editor_str)
+
+def get_commit_message_template() -> str:
+    initial_text = """\
+# Please enter the commit message for your changes. Lines starting
+# with '#' will be ignored, and an empty message aborts the commit.
+#
+"""
```

Comment:
Not sure if this is just a personal nit, but should we pull this variable out of the function? Since Python is an indented language, I've always found the multi-line string dedents to throw off the flow of a function definition. Could totally just be a personal quibble though 😅

Reply:
Done! Agreed it's slightly gross as-written -- I was waiting to see if there was consensus around this being worth pursuing before cleaning it up.